### PR TITLE
Network socket API syscalls

### PR DESCRIPTION
**Pull request by tyrantlizard77**
Add system calls for socket APIs.

The system call for close() is not completely thread-safe, but will require an overhaul on how kernel objects are validated which will be tackled in another PR.

### CODE REVIEW

**Review by jukkar — CHANGES_REQUESTED** [review]
Looks quite simple actually, minor tweak needed to avoid use of magic constants in socket code.

File: subsys/net/lib/sockets/sockets.c [hunk 1]
```diff
@@ -470,10 +585,56 @@ ssize_t zsock_recvfrom(int sock, void *buf, size_t max_len, int flags,
...
+	socklen_t addrlen_copy;
+	socklen_t *addrlen_ptr = (socklen_t *)addrlen_param;
+	ssize_t ret;
+
+	if (Z_SYSCALL_OBJ(sock, K_OBJ_NET_CONTEXT)) {
+		errno = EBADF;
+		return -1;
+	}
+
+	struct net_context *ctx = INT_TO_POINTER(sock);
```
Comment (jukkar) [inline subsys/net/lib/sockets/sockets.c]: Shouldn't these variables declared at the beginning of the func?

Comment (andrewboie) [reply subsys/net/lib/sockets/sockets.c]: Can you show me where this is mandated in our coding style? I can't find any prohibition against it and checkpatch is AFAIK happy.

**Review by jukkar — APPROVED** [review]
+1 to network changes

File: kernel/include/syscall_handler.h [hunk 2]
```diff
@@ -125,6 +125,19 @@ extern void _thread_perms_all_clear(struct k_thread *thread);
...
+ * Initialize and reset permissions to only access by the caller
+ *
+ * Intended for scenarios where objects are fetched from slab pools
+ * and may have had different permissions set during prior usage.
+ *
+ * The object will be marked as initialized and the calling thread
+ * granted access to it.
+ *
+ * @param object Address of the kernel object
+ */
```
Comment (andyross) [inline kernel/include/syscall_handler.h]: This one is a little confusing.   I'm gathering that the idea is to take an object that you already have and "reinitialize" it so you don't have to go through the pair of k_obj_init()/uninit() calls?  If so, maybe document it like that?

But then I'm looking at k_object_init() and noticing that it doesn't reset the perms like this function does, and wondering whether that should have happened in the first place?

Basically the need for this makes me worried that maybe we have a messy API for object initialization state and should probably clean it up instead of adding new pseudoinitialization callls. :)

File: kernel/userspace.c [hunk 3]
```diff
@@ -17,6 +17,9 @@
 #include <device.h>
 #include <init.h>
 #include <logging/sys_log.h>
+#ifdef CONFIG_NETWORKING
```
Comment (andyross) [inline kernel/userspace.c]: You #if-out this include, but I notice that none of your other changes to this file need to live underneath the same #if.  So either this #if isn't necessary or you missed a spot that is going to break when networking is disabled.

File: tests/net/socket/getaddrinfo/src/main.c [hunk 4]
```diff
@@ -41,9 +41,11 @@ void test_getaddrinfo_no_host(void)
 
 void test_main(void)
 {
+	k_thread_system_pool_assign(k_current_get());
```
Comment (andyross) [inline tests/net/socket/getaddrinfo/src/main.c]: Does this do anything?  ztest runs the suite in a separate thread, changing the malloc pool for this thread is a noop, right?

**Review by andyross — CHANGES_REQUESTED** [review]
Nitpicking mostly.  The boilerplate validation looks fine but I'm not 100% sure I know what to look for.

The -1 is because the _recycle() call is giving me the jeebies that we've got some holes in our kobject lifecycle that should be plugged, but maybe that's possible with docs or explanation.

File: kernel/userspace.c (hunk 3 above)
Comment (andyross) [reply kernel/userspace.c]: Or the third option: the header wasn't needed to begin with.

File: tests/net/socket/getaddrinfo/src/main.c (hunk 4 above)
Comment (andrewboie) [reply tests/net/socket/getaddrinfo/src/main.c]: yes, it does.
It assigns the system heap to the thread resource pool of the main thread.
Child threads, including all the ztest threads, inherit this assignment.

File: kernel/userspace.c (hunk 3 above)
Comment (andrewboie) [reply kernel/userspace.c]: This is quite necessary. The auto-generated otype-to-size.h which gets included by this file needs to know the size of a struct net_context if networking is enabled.

I don't include this without an ifdef because you get build errors depending on kernel configuration if you pull this in unconditionally.

File: kernel/include/syscall_handler.h (hunk 2 above)
Comment (andrewboie) [reply kernel/include/syscall_handler.h]: Starting to write a lengthy reply, but probably better to just amend our documentation to explain the distinction between k_object_init() and k_object_recycle() as they are used for different purposes, so I'll do that instead.

File: kernel/userspace.c (hunk 3 above)
Comment (andyross) [reply kernel/userspace.c]: Then maybe the auto-generated header should include the proper files for its own dependencies?  Or at least document that above the include?  I complain because I'd totally see this kind of thing in the future and rip it out and then waste time trying to figure out what breaks.

Comment (andrewboie) [reply kernel/userspace.c]: It's a generated switch body:

static size_t obj_size_get(enum k_objects otype)
{
	switch (otype) {
#include <otype-to-size.h>
	default:
		return sizeof(struct device);
	}
}

It contains:

case K_OBJ_ALERT: return sizeof(struct k_alert);
case K_OBJ_MSGQ: return sizeof(struct k_msgq);
case K_OBJ_MUTEX: return sizeof(struct k_mutex);
case K_OBJ_PIPE: return sizeof(struct k_pipe);
case K_OBJ_QUEUE: return sizeof(struct k_queue);
case K_OBJ_POLL_SIGNAL: return sizeof(struct k_poll_signal);
case K_OBJ_SEM: return sizeof(struct k_sem);
case K_OBJ_STACK: return sizeof(struct k_stack);
case K_OBJ_THREAD: return sizeof(struct k_thread);
case K_OBJ_TIMER: return sizeof(struct k_timer);
#ifdef CONFIG_NETWORKING
case K_OBJ_NET_CONTEXT: return sizeof(struct net_context);
#endif

@lpereira added this to reduce boilerplate when adding new kernel objects.

Comment (andrewboie) [reply kernel/userspace.c]: >  Or at least document that above the include? 

Can definitely add some comments explaining what's going on.

File: subsys/net/lib/sockets/sockets.c [hunk 5]
```diff
@@ -84,14 +85,34 @@ int zsock_socket(int family, int type, int proto)
...
+	return _impl_zsock_socket(family, type, proto);
+}
+#endif /* CONFIG_USERSPACE */
+
+int _impl_zsock_close(int sock)
 {
 	struct net_context *ctx = INT_TO_POINTER(sock);
 
+#ifdef CONFIG_USERSPACE
+	_k_object_uninit(ctx);
```
[the remark this answers is not in the document]
Comment (andrewboie) [reply subsys/net/lib/sockets/sockets.c]: No, we don't want any initialization checks to succeed while the object is being disposed of

File: kernel/userspace.c [hunk 6]
```diff
@@ -462,6 +465,17 @@ void _k_object_init(void *object)
 	ko->flags |= K_OBJ_FLAG_INITIALIZED;
 }
 
+void _k_object_recycle(void *object)
+{
+	struct _k_object *ko = _k_object_find(object);
```
[the remark this answers is not in the document]
Comment (andrewboie) [reply kernel/userspace.c]: We do not.
Supervisor mode is allowed to declare and use un-tracked kernel objects.

[the remark this answers is not in the document]
Comment (pizi-nordic) [reply kernel/userspace.c]: Ok.

File: subsys/net/lib/sockets/sockets.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply subsys/net/lib/sockets/sockets.c]: Ok

File: kernel/userspace.c [hunk 7]
```diff
@@ -462,6 +465,17 @@ void _k_object_init(void *object)
...
+
+	if (ko) {
+		memset(ko->perms, 0, CONFIG_MAX_THREAD_BYTES);
```
[the remark this answers is not in the document]
Comment (pfalcon) [reply kernel/userspace.c]: > IMHO memset(ko->perms, 0, sizeof(ko->perms)) is a bit safer from maintenance perspective.

Yes, but maybe sizeof(ko->perms) != CONFIG_MAX_THREAD_BYTES? I definitely don't immediately see a connection between "max thread bytes" and "permissions" ;-).

File: include/net/socket.h [hunk 8]
```diff
@@ -97,26 +97,68 @@ struct zsock_addrinfo {
...
-		       struct sockaddr *src_addr, socklen_t *addrlen);
-int zsock_fcntl(int sock, int cmd, int flags);
-int zsock_poll(struct zsock_pollfd *fds, int nfds, int timeout);
+/*
+ * NOTE NOTE NOTE NOTE
+ *
+ * System calls have been added at the socket API level so that we can
+ * provide user mode access to the network stack fairly quickly.
+ *
+ * In the fullness of time, the system call layer should be moved much
```
Comment (pfalcon) [inline include/net/socket.h]: I wouldn't leave this note in the version to merge. Nor I'm sure the statement ("In the fullness of time, the system call layer should be moved much further down the stack") is universally right. How down is "much further down"?, etc. Better not leave such prophecies in the source code. Could (should?) go into the commit message though.

File: include/net/socket.h [hunk 9]
```diff
@@ -97,26 +97,68 @@ struct zsock_addrinfo {
...
-int zsock_getaddrinfo(const char *host, const char *service,
+
+__syscall int zsock_inet_pton(sa_family_t family, const char *src, void *dst);
+
+__syscall int z_zsock_getaddrinfo_internal(const char *host,
+					   const char *service,
+					   const struct zsock_addrinfo *hints,
+					   struct zsock_addrinfo *res);
+
+int zsock_getaddrinfo(const char *host,	const char *service,
```
Comment (pfalcon) [inline include/net/socket.h]: There's something wrong with whitespace after "*host,".

File: include/net/socket.h [hunk 10]
```diff
@@ -97,26 +97,68 @@ struct zsock_addrinfo {
...
 		     void *optval, socklen_t *optlen);
+
 int zsock_setsockopt(int sock, int level, int optname,
 		     const void *optval, socklen_t optlen);
-int zsock_inet_pton(sa_family_t family, const char *src, void *dst);
-int zsock_getaddrinfo(const char *host, const char *service,
+
+__syscall int zsock_inet_pton(sa_family_t family, const char *src, void *dst);
+
+__syscall int z_zsock_getaddrinfo_internal(const char *host,
```
Comment (pfalcon) [inline include/net/socket.h]: What is significance of "z_" prefix to the already existing "zsock_" prefix?

File: subsys/net/lib/sockets/getaddrinfo.c [hunk 11]
```diff
@@ -164,6 +165,46 @@ static int zsock_getaddrinfo_internal(const char *host, const char *service,
...
 }
 
+#ifdef CONFIG_USERSPACE
+Z_SYSCALL_HANDLER(z_zsock_getaddrinfo_internal, host, service, hints, res)
+{
+	struct zsock_addrinfo hints_copy;
+	char *host_copy = NULL, *service_copy = NULL;
+	u32_t ret;
+
+	if (hints) {
```
Comment (pfalcon) [inline subsys/net/lib/sockets/getaddrinfo.c]: This may be minor, but what's the ordering pattern of processing params here? I'd think they should be processed in the original order, or maybe group by in/out-ness, original order in each group. Here, the order appears random. (I pat attention to such things because if I had to do that, I'd copy-pasted existing code, and existing code doesn't try to teach to do it in some consistent pattern.)

File: include/net/socket.h (hunk 8 above)
Comment (pfalcon) [reply include/net/socket.h]: (Of course, if you mean "syscalls" like getaddrinfo and inet_pton - yeah, those could live in userspace completely. The core stuff like socket/send/recv are pretty much kernel stuff IMHO. Btw, chosen TLS support implementation complicates that too.)

File: subsys/net/lib/sockets/sockets.c [hunk 12]
```diff
@@ -84,14 +85,34 @@ int zsock_socket(int family, int type, int proto)
 	/* recv_q and accept_q are in union */
 	k_fifo_init(&ctx->recv_q);
 
+#ifdef CONFIG_USERSPACE
+	/* Set net context object as initialized and grant access to the
+	 * calling thread (and only the calling thread)
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: > (and only the calling thread)

I love that - no I/O resource sharing between threads, kids! ;-)

File: subsys/net/lib/sockets/sockets.c [hunk 13]
```diff
@@ -202,12 +271,29 @@ int zsock_listen(int sock, int backlog)
...
+#endif /* CONFIG_USERSPACE */
+
+int _impl_zsock_accept(int sock, struct sockaddr *addr, socklen_t *addrlen)
 {
 	struct net_context *parent = INT_TO_POINTER(sock);
 
 	struct net_context *ctx = k_fifo_get(&parent->accept_q, K_FOREVER);
 
+#ifdef CONFIG_USERSPACE
+	_k_object_init(ctx);
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: Well, ctx is gotten from fifo, but it gets into fifo from the native stack, which net_context_get's it. So, I don't see why it wouldn't be _k_object_recycle() here, just the same as for socket(). If you see, that should go into code comments, because many will be confused going forward.

File: subsys/net/lib/sockets/sockets.c [hunk 14]
```diff
@@ -229,13 +315,40 @@ int zsock_accept(int sock, struct sockaddr *addr, socklen_t *addrlen)
...
-	return zsock_sendto(sock, buf, len, flags, NULL, 0);
+	socklen_t addrlen_copy;
+	int ret;
+
+	if (Z_SYSCALL_OBJ(sock, K_OBJ_NET_CONTEXT)) {
+		errno = EBADF;
+		return -1;
+	}
+
+	Z_OOPS(z_user_from_copy(&addrlen_copy, (void *)addrlen,
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: Wait, what happens here? Casting addren to pointer?..

File: subsys/net/lib/sockets/sockets.c [hunk 15]
```diff
@@ -229,13 +315,40 @@ int zsock_accept(int sock, struct sockaddr *addr, socklen_t *addrlen)
...
+
+	if (Z_SYSCALL_MEMORY_WRITE(addr, addrlen_copy)) {
+		errno = EFAULT;
+		return -1;
+	}
+
+	ret = _impl_zsock_accept(sock, (struct sockaddr *)addr, &addrlen_copy);
+
+	if (ret >= 0 &&
+	    z_user_to_copy((void *)addrlen, &addrlen_copy,
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: Funny copy again.

File: subsys/net/lib/sockets/sockets.c (hunk 14 above)
Comment (pfalcon) [reply subsys/net/lib/sockets/sockets.c]: Ah damn, nevermind, it's addrlen_copy. Let me remove these noise comments...

Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: OTOH, I don't understand why we allow kernel to write directly into addr, and don't allow into addrlen...

File: subsys/net/lib/sockets/sockets.c [hunk 16]
```diff
@@ -285,6 +398,28 @@ ssize_t zsock_sendto(int sock, const void *buf, size_t len, int flags,
...
+	struct sockaddr_storage dest_addr_copy;
+
+	if (Z_SYSCALL_OBJ(sock, K_OBJ_NET_CONTEXT)) {
+		errno = EBADF;
+		return -1;
+	}
+
+	Z_OOPS(Z_SYSCALL_MEMORY_READ(buf, len));
+	if (dest_addr) {
+		Z_OOPS(z_user_from_copy(&dest_addr_copy, (void *)dest_addr,
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: How do we handle here (and in other places) a case when addrlen > sizeof(dest_addr_copy) ?

File: subsys/net/lib/sockets/sockets.c [hunk 17]
```diff
@@ -470,10 +600,58 @@ ssize_t zsock_recvfrom(int sock, void *buf, size_t max_len, int flags,
...
+
+	if (Z_SYSCALL_OBJ(sock, K_OBJ_NET_CONTEXT)) {
+		errno = EBADF;
+		return -1;
+	}
+
+	ctx = INT_TO_POINTER(sock);
+	sock_type = net_context_get_type(ctx);
+
+	if (sock_type != SOCK_DGRAM && sock_type != SOCK_STREAM) {
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: Why suddenly validate this for recvfrom()? (Not validated for anything else, e.g. sendto).

**Comment by pfalcon** [comment]
Tried to eyeball thoroughly, added a bunch of suggestions and questions.

File: subsys/net/lib/sockets/sockets.c (hunk 12 above)
Comment (andrewboie) [reply subsys/net/lib/sockets/sockets.c]: That's not how it works.

Additional threads may be granted access to the context object with k_object_access_grant(). This is just the default policy for newly acquired objects.

File: include/net/socket.h (hunk 10 above)
Comment (andrewboie) [reply include/net/socket.h]: z_ naming convention is the new way of denoting private APIs.

File: subsys/net/lib/sockets/sockets.c (hunk 16 above)
Comment (andrewboie) [reply subsys/net/lib/sockets/sockets.c]: if z_user_from_copy() fails the calling thread is killed.

File: subsys/net/lib/sockets/sockets.c (hunk 17 above)
Comment (andrewboie) [reply subsys/net/lib/sockets/sockets.c]: because the recvfrom implementation doesn't validate properly, it only has an __ASSERT

File: subsys/net/lib/sockets/sockets.c (hunk 14 above)
Comment (andrewboie) [reply subsys/net/lib/sockets/sockets.c]: addrlen is an input/output parameter whose contents must be copied to prevent TOCTOU issues.

File: subsys/net/lib/sockets/getaddrinfo.c (hunk 11 above)
Comment (andrewboie) [reply subsys/net/lib/sockets/getaddrinfo.c]: There's no particular ordering requirement, other than to validate pointers and copy any containing information before using it.

File: kernel/include/syscall_handler.h (hunk 2 above)
Comment (andrewboie) [reply kernel/include/syscall_handler.h]: The doc has been expanded a bit.

Basically, for statically defined objects, you don't want this. If I declare a thread object toplevel, the permissions on it should persist even if the thread object is re-used.

However, for objects that are drawn from a pool, it needs it's permissions freshened up each time a pool member is returned to the user. It's past permission state needs to be wiped clean.

File: subsys/net/lib/sockets/sockets.c (hunk 16 above)
Comment (andrewboie) [reply subsys/net/lib/sockets/sockets.c]: Oh I see it, yes this needs to be accounted for.

**Review by andyross — APPROVED** [review]
OK, I get it.  I'll leave in a vestigial whine about naming: we have "recycle" which changes one part of the object state and "un/init" which changes another, and they don't really speak to each other.   Going from the existing API to a mental model of a kobject lifecycle seems needlessly complicated still.  But the underlying data isn't very complicated, so maybe that's not a big deal.
